R/glmCommonFunctions.R

```diff
@@ -0,0 +1,207 @@
+  } else {
+
+    if (nullModel) {
+      for (i in seq_along(modelTerms)) {
+        nuisance <- modelTerms[[i]][["isNuisance"]]
+        if (!is.null(nuisance) && nuisance) {
+          term <- modelTerms[[i]][["components"]]
+          if (length(term) == 1)
+            rightTerms <- c(rightTerms, term)
+          else
+            rightTerms <- c(rightTerms, paste(term, collapse = ":"))
+        }
+      }
+
+    } else {
+      for (i in seq_along(modelTerms)) {
+        term <- modelTerms[[i]][["components"]]
+        if (length(term) == 1)
+          rightTerms <- c(rightTerms, term)
+        else
+          rightTerms <- c(rightTerms, paste(term, collapse = ":"))
+      }
+    }
+    f <- formula(paste(leftTerm, "~", paste(rightTerms, collapse = "+")))
+  }
+  return(f)
+}
+
+
+.glmComputeModel <- function(jaspResults, dataset, options) {
+  if (!is.null(jaspResults[["glmModels"]]))
+    return(jaspResults[["glmModels"]]$object)
+
+  # make formulas for the full model and the null model
+  ff <- .createGLMFormula(options, nullModel = FALSE)
+  nf <- .createGLMFormula(options, nullModel = TRUE)
+  # specify family and link
+  if (options$family == "bernoulli")
+    family <- "binomial"
+  else
+    family <- options$family
+  familyLink <- eval(call(family, link = options$link))
+  # compute full and null models
+  if (options$weights == "") {
+    fullModel <- stats::glm(ff, family = familyLink, data = dataset, weights = NULL)
+    nullModel <- stats::glm(nf, family = familyLink, data = dataset, weights = NULL)
+  } else {
+    fullModel <- stats::glm(ff,
+                            family = familyLink,
+                            data = dataset,
+                            weights = get(options$weights))
+    nullModel <- stats::glm(nf,
+                            family = familyLink,
+                            data = dataset,
+                            weights = get(options$weights))
+  }
+
+  glmModels <- list("nullModel" = nullModel,
+                    "fullModel" = fullModel)
+  # combine both models
+  jaspResults[["glmModels"]] <- createJaspState()
+  jaspResults[["glmModels"]]$dependOn(optionsFromObject = jaspResults[["modelSummary"]])
+  jaspResults[["glmModels"]]$object <- glmModels
+
+  return(glmModels)
+
+}
+
+.hasNuisance <- function(options) {
+  return(any(sapply(options$modelTerms, function(x) x[["isNuisance"]])))
+}
+
+.glmCreatePlotPlaceholder <- function(container, index, title) {
+  jaspPlot            <- createJaspPlot(title = title)
+  jaspPlot$status     <- "running"
+  container[[index]]  <- jaspPlot
+  return()
+}
+
+
+.glmStdResidCompute <- function(model, residType) {
+  if (residType == "deviance") {
+    stdResid <- rstandard(model)
+  }
+  else if (residType == "Pearson") {
+    phiEst <- summary(model)[["dispersion"]]
+    resid <- resid(model, type="pearson")
+    stdResid <- resid / sqrt(phiEst * (1 - hatvalues(model)))
+  }
+  else if (residType == "quantile") {
+    set.seed(123)
```

Comment:
Why do you set see here?

Reply:
Because the next line `resid <- statmod::qresid(model)` requires setting a seed to make reproducible results (i.e. quantile residuals), which are drawn from some probability distributions.

Reply:
ok, it's surprising to me that computing the quantile residuals requires randomization, but ok...

Then I would advice exposing this to the user, using the `SetSeed` QML component + using the `jaspBase::.setSeedJASP` function to set the seed.

Reply:
Only one `SetSeed` component in the analysis.

Here is an example of using it in QML: https://github.com/jasp-stats/jaspMixedModels/blob/32f26c6c4cdb39ac685c0326208946f8794d42b9/inst/qml/common/MixedModelsBOptions.qml#L107

And then using it in R: https://github.com/jasp-stats/jaspMixedModels/blob/e2ae980443f5731c3fb2547049d493d66c67d6e0/R/MixedModelsCommon.R#L442